{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "54b59b91",
   "metadata": {},
   "source": [
    "COUNTING WORDS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4655101d",
   "metadata": {},
   "source": [
    "before diving into our first algorithm , lets run a very nasic one first"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b1e3691",
   "metadata": {},
   "source": [
    "Use what you have learned to print the length of the Vibrio cholerae ori, which is found here. You can see more directions when you click \"Start Solving\"; you have as many attempts to solve the problem as you like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0a0f06fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "540\n"
     ]
    }
   ],
   "source": [
    "ori = \"ATCAATGATCAACGTAAGCTTCTAAGCATGATCAAGGTGCTCACACAGTTTATCCACAACCTGAGTGGATGACATCAAGATAGGTCGTTGTATCTCCTTCCTCTCGTACTCTCATGACCACGGAAAGATGATCAAGAGAGGATGATTTCTTGGCCATATCGCAATGAATACTTGTGACTTGTGCTTCCAATTGACATCTTCAGCGCCATATTGCGCTGGCCAAGGTGACGGAGCGGGATTACGAAAGCATGATCATGGCTGTTGTTCTGTTTATCTTGTTTTGACTGAGACTTGTTAGGATAGACGGTTTTTCATCACTGACTAGCCAAAGCCTTACTCTGCCTGACATCGACCGTAAATTGATAATGAATTTACATGCTTCCGCGACGATTTACCTCTTGATCATCGATCCGATTGAAGATCTTCAATTGTTAATTCTCTTGCCTCGACTCATAGCCATGATGAGCTCTTGATCATGTTTCCTTAACCCTCTATTTTTTACGGAAGAATGATCAAGCTGCTGCTCTTGATCATCGTTTC\"\n",
    "\n",
    "print(len(ori))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "108d1763",
   "metadata": {},
   "source": [
    "this is the nucleotide sequence appearing in the ori region of the Vibrio Cholerae"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4026e63b",
   "metadata": {},
   "source": [
    "# COUNTING WORDS\n",
    "\n",
    "\n",
    "Operating under the assumption that DNA is a language of its own, let’s borrow Legrand’s method and see if we can find any surprisingly frequent “words” within the ori of Vibrio cholerae. We have added reason to look for frequent words in the ori because for various biological processes, certain nucleotide strings appear surprisingly often in small regions of the genome. This is often because certain proteins can only bind to DNA if a specific string of nucleotides is present, and if there are more occurrences of the string, then it is more likely that binding will successfully occur. (It is also less likely that a mutation will disrupt the binding process.)\n",
    "\n",
    "For example, \"ACTAT\" is a surprisingly frequent substring of\n",
    "\n",
    "\"ACAACTATGCATACTATCGGGAACTATCCT\".\n",
    "\n",
    "We use the term k-mer for a string of length k and define PatternCount(Pattern, Text) as the number of times that a k-mer Pattern appears as a substring of a string Text. Following the above example,\n",
    "\n",
    "PatternCount(\"ACTAT\", \"ACAACTATGCATACTATCGGGAACTATCCT\") = 3.\n",
    "\n",
    "Note that PatternCount(\"ATA\", \"CGATATATCCATAG\") is equal to 3 (not 2) since we should account for overlapping occurrences of Pattern in Text."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd6dc40",
   "metadata": {},
   "source": [
    "Write the program to count the number of patter or k-mer in the fiven text or genome\n",
    "\n",
    "Sample Input:\n",
    "GCGCG;;;\n",
    "GCG\n",
    "\n",
    "Sample Output:\n",
    "2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1700dc83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#def PatternCount(Text, Pattern):\n",
    "\n",
    "def PatternCount(text , pattern): \n",
    "    count = 0\n",
    "    for i in range(len(text)-len(pattern)+1) :\n",
    "        if text[i:i+len(pattern)] == pattern :\n",
    "            count = count + 1\n",
    "    return count\n",
    "\n",
    "# PatternCount(\"GCGCG\",\"GCG\")\n",
    "\n",
    "PatternCount(\"AAAAAAA\",\"AAA\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a54a6e97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#def PatternCount(Text, Pattern):\n",
    "\n",
    "def PatternCount(text , pattern): \n",
    "    count = 0\n",
    "    for i in range(len(text)-len(pattern)+1) :\n",
    "        if text[i:i+len(pattern)] == pattern :\n",
    "            count = count + 1\n",
    "    return count\n",
    "\n",
    "PatternCount(\"GACCATCAAAACTGATAAACTACTTAAAAATCAGT\",\"AAA\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0947295a",
   "metadata": {},
   "source": [
    "this function helps us to count the number of occurence of the given k-mer or pattern from the gene or terxt provided"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
